core/src/main/scala/org/apache/spark/deploy/ExternalShuffleService.scala

```diff
@@ -56,11 +60,55 @@ class ExternalShuffleService(sparkConf: SparkConf, securityManager: SecurityMana
 
   private var server: TransportServer = _
 
+  private final val  MAX_DIR_CREATION_ATTEMPTS = 10
+
   private val shuffleServiceSource = new ExternalShuffleServiceSource
 
+  protected def createDirectory(root: String, name: String): File = {
+    var attempts = 0
+    val maxAttempts = MAX_DIR_CREATION_ATTEMPTS
+    var dir: File = null
+    while (dir == null) {
+      attempts += 1
+      if (attempts > maxAttempts) {
+        throw new IOException("Failed to create a temp directory (under " + root + ") after " +
+          maxAttempts + " attempts!")
+      }
+      try {
+        dir = new File(root, "registeredExecutors")
+        if (!dir.exists() && !dir.mkdirs()) {
+          dir = null
+        }
+      } catch { case e: SecurityException => dir = null; }
+    }
+    logInfo(s"registeredExecutorsDb path is ${dir.getAbsolutePath}")
+    new File(dir.getAbsolutePath, name)
+  }
+
+  protected def initRegisteredExecutorsDB(dbName: String): File = {
+    val localDirs = sparkConf.get("spark.local.dir", "").split(",")
+    if (localDirs.length >= 1 && !"".equals(localDirs(0))) {
+      createDirectory(localDirs(0), dbName)
```

Comment:
any reason to only use `localDirs(0)` instead of checking all localDirs?  I worry that you might get another local dir pre-pended or something during a configuration change + restart

also it seems you're creating a path like "[local-dir]/registeredExecutors/registeredExecutors.ldb", any reason for the extra level?

Reply:
Hi, @squito , as you agreed that it certainly should be possible to do it. 
`the one which comes to mind is, what happens if an application is stopped while the external shuffle service is down?  In yarn, we rely on being told the application was stopped even after the NM comes back.`
Now , It  can  leave an entry in the DB forever when some time like above.   As you said that  Maybe this is rare enough and low-impact enough ,  but at least worth thinking through and documenting. So I think  we can add some core to remove  the entry  with WorkDirCleanup  when set #spark.worker.cleanup.enabled = true in standalone mode. can you have any good idea ? 

This commit uses  `localDirs(0)` instead of checking all localDirs to make sure it's  a same path to be used by DB  and  make sure initRegisteredExecutorsDB  to work , `localDirs(0) ` is just to 
be used  for DB  instead of  additional set. 

Creating a path like "[local-dir]/registeredExecutors/registeredExecutors.ldb" is just to make it look clearly .

Reply:
yes I think WorkDirCleanup maybe just what we need to ensure things get cleaned up, good idea.

I understand wanting to use a consistent directory, but like I said I'm worried about restarts after configuration changes (maybe not a concern in a standalone mode?  does it always require a total restart?)  You could do something like what was done in the original patch for yarn, to check all the dirs, but fallback to dir[0] (that code has since changed to take advantage of other yarn features for recovery):

https://github.com/apache/spark/blob/708036c1de52d674ceff30ac465e1dcedeb8dde8/network/yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java#L192-L200